fili-core/src/main/java/com/yahoo/bard/webservice/data/filterbuilders/DruidInFilterBuilder.java

```diff
@@ -0,0 +1,83 @@
+// Copyright 2016 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.data.filterbuilders;
+
+import com.yahoo.bard.webservice.data.dimension.Dimension;
+import com.yahoo.bard.webservice.data.dimension.DimensionRowNotFoundException;
+import com.yahoo.bard.webservice.druid.model.filter.AndFilter;
+import com.yahoo.bard.webservice.druid.model.filter.Filter;
+import com.yahoo.bard.webservice.druid.model.filter.InFilter;
+import com.yahoo.bard.webservice.druid.model.filter.NotFilter;
+import com.yahoo.bard.webservice.web.ApiFilter;
+import com.yahoo.bard.webservice.web.FilterOperation;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A DruidInFilterBuilder builds a conjunction of in-filters for each Dimension, where each in-filter corresponds to a
+ * filter term. So, the filter terms on dimension {@code category}:
+ * <p>
+ * {@code category|id-in[finance,sports],category|desc-contains[baseball]}
+ * <p>
+ * are translated into:
+ * <pre>
+ *     {@code
+ *     {
+ *         "type": "in",
+ *         "dimension": "category",
+ *         "values": ["finance", "sports"]
+ *     }
+ *     }
+ * </pre>
+ * Each filter term is resolved independently of the other filter terms.
+ */
+public class DruidInFilterBuilder extends ConjunctionDruidFilterBuilder {
+    private static final Logger LOG = LoggerFactory.getLogger(DruidInFilterBuilder.class);
+
+    @Override
+    protected Filter buildDimensionFilter(Dimension dimension, Set<ApiFilter> filters)
+            throws DimensionRowNotFoundException {
+        LOG.trace("Building dimension filter using dimension: {} \n\n and set of filter: {}", dimension, filters);
+
+        List<Filter> inFilters = new ArrayList<>(); // A list with at most 2 in-filters(positive & negative)
+        Set<String> positiveInValues = new HashSet<>(); // contains values for positive in-filter
+        Set<String> negativeInValues = new HashSet<>(); // contains values for negative in-filter
+
+        for (ApiFilter filter : filters) {
```

Comment:
Taking another look at this, it looks like all of these filters are from
the same dimension.

So why are we resolving each filter individually? It seems to me like this would
be extremely inefficient in many realword cases (especially if you have 
autogenerated filters, like filters for row security).

Also, it's wrong.

For example, suppose we have the filters:

```java
    dimension|name-contains[e],dimension|id-in[1,2,3]
```

Furthermore, suppose this dimension has 10,000+ rows, and every name is of the
form `e-somethingorother`.

In the example above, our end result should be 
`in(1, 2, 3)`. However, we never actually trim any results out of 
`positiveInValues`. So instead the `inFilter` will contain 10,000+ values, even
though it should only have 3.

However, even if we fix that bug, this is still does not scale at all.
This filter shouldn't actually be that bad, assuming whatever is
backing your SearchProvider (i.e. Lucene) is at all smart about how it handles
`in` vs.  `contains`, which Lucene probably is considering its maturity.
The way this is currently implemented, we're not taking advantage of 
*any* intelligence that may or may not exist in the SearchProvider 
implementation, because we're executing each filter clause individually. 

And even if the SearchProvider is dumb as a box of rocks, this is still not 
an ideal implementation, because we're allocating a set with 10,000+ 
values, copying 10,000 numbers into it, and then later trimming it down to 3,
instead of just getting 3 values back from the SearchProvider.

What we really should be doing I think, is just sending all the filters down to
the SearchProvider at once if we have at least one positive filter.

The only time we should be doing anything remotely like this is if all the
filters are negative. In that case, we would have to send each filter down
individually, because we have no way of encoding top-level ors, so we can't
just pull out the common negative and send down positive versions of all the
filters.

So if we have:

```java
    dimension|name-contains[e],dimension|id-in[1,2,3],dimension|id-notin[1]
```

Then we should just send this whole set down to the SearchProvider at once.

If we have the filter:

```java
    dimension|id-notin[1,2,3],dimension|name-notin[e-1234],dimension|id-notin[5]
```
Then yes, we'll need to send each filter down individually, because this is 
equivalent to

```
NOT(dimension|id-in[1,2,3] OR dimension|name-in[e-1234] OR dimension|id-in[5])
```

and we have no way of encoding the OR's.

Reply:
To specify the application Andrew's point, the algorithm we should be following is:
a) group collect positive and negative constraints
b) let the search provider resolve all the positive constraints, but a Druid in-filter out of those.
c) let the search provider resolve all the negative constraints, but a druid in-filter out of those and encapsulate it in a not filter
d) And those results together.